clang/include/clang/AST/PreorderAST.h

```diff
@@ -0,0 +1,165 @@
+          // way to indicate the absence of a Constant. So we have a boolean
+          // HasConstant to indicate whether a node has a Constant or not.
+          Constant = GetConstVal(0);
+        }
+
+      llvm::APSInt GetConstVal(unsigned Val) {
+        return llvm::APSInt(llvm::APInt(Ctx.getTypeSize(Ctx.IntTy), Val));
+      }
+
+      void AddVar(std::string Name) {
+        // We initialize the count of each variable in a node to 1.
+        Variables.push_back(VarTy {Name, GetConstVal(1)});
+      }
+
+      // Is the operator commutative and associative?
+      bool IsOpCommutativeAndAssociative() {
+        return Opcode == BO_Add ||
+               Opcode == BO_Mul;
+      }
+    };
+
+  private:
+    ASTContext &Ctx;
+    Lexicographic Lex;
+    llvm::raw_ostream &OS;
+    bool Error;
+    ASTNode *AST;
+
+    // Create a preorder AST from the expression E.
+    // @param[in] N is the current node of the AST.
+    // @param[in] E is the sub expression which needs to be added to N.
+    // @param[in] Parent is the Parent node for N.
+    void Create(ASTNode *N, Expr *E, ASTNode *Parent = nullptr);
+
+    // Normalize the input expression through a series of transforms on the
+    // preorder AST. The Error field is set if an error is encountered during
+    // transformation of the AST.
+    // @param[in] N is the root of the AST.
+    void Normalize(ASTNode *N);
+
+    // Sort the variables in a node of the AST. The Error field is set if an
+    // error is encountered during transformation of the AST.
+    // @param[in] N is the root of the AST.
+    void Sort(ASTNode *N);
+
+    // Check if the two ASTs N1 and N2 are equal.
+    // @param[in] N1 is the first AST.
+    // @param[in] N2 is the second AST.
+    // @return Returns a boolean indicating whether N1 and N2 are equal.
+    bool IsEqual(ASTNode *N1, ASTNode *N2);
+
+    // Cleanup the memory consumed by the AST.
+    // @param[in] N is the root node of the AST.
+    void Cleanup(ASTNode *N);
+
+    // Set Error in case an error occurs during transformation of the AST.
+    // @param[in] Err is the value to be set for the Error field.
+    void SetError(bool Err) {
+      Error = Err;
+    }
+
+    // Print the preorder AST.
+    // @param[in] N is the root node of the AST.
+    void PrettyPrint(ASTNode *N);
+
+    // A DeclRefExpr can be a reference either to an array subscript (in which
+    // case it is wrapped around a ArrayToPointerDecay cast) or to a pointer
+    // dereference (in which case it is wrapped around an LValueToRValue cast).
+    // @param[in] An expression E.
+    // @return Returns a DeclRefExpr if E is a DeclRefExpr, otherwise nullptr.
+    DeclRefExpr *GetDeclOperand(Expr *E);
+
+  public:
+    PreorderAST(ASTContext &Ctx, Expr *E) :
+      Ctx(Ctx), Lex(Lexicographic(Ctx, nullptr)),
+      OS(llvm::outs()), Error(false) {
+
+      AST = new ASTNode(Ctx);
+      Create(AST, E);
+      Normalize(AST);
```

Comment:
It would be better to not have the constructor normalize expressions.  That makes code harder to read and understand because there's an implicit operation at construction of a PreorderAST.